packages/javascript-sdk/src/logger/index.ts

```diff
@@ -0,0 +1,47 @@
+import Config from '../config/index';
+
+interface LoggerFunctions {
+  warnFn?: (...msgs: unknown[]) => void;
+  errorFn?: (...msgs: unknown[]) => void;
+  logFn?: (...msgs: unknown[]) => void;
+}
+
+function customLogger() {
```

Comment:
So, this is a pretty new convention we're setting here. In the case of the `TokenStore`, we passed the custom store object in the `Config.set` method. Since this is new, can we do a quick cost-benefit of adding this new convention?

Reply:
So this was my thought process.

I started with an abstract class, because it made sense that we didn't need to instantiate any class and it reduced the footprint of changes in the SDK.

Then when we suggested adding a custom logger, an abstract class didn't really work unless we modify the config level code. Once we pass in a custom logger, theres no real way of passing arguments through an abstract config because its static, and it doesn't get invoked. 

I didn't really know if we wanted to expand Config beyond the general "ForgeRock" AM configuration settings and I felt logging wasn't AM Configuration but more preference of the consumer.

The downside being there needs to be an explicit call of the FRLogger in this case to configure logging, but its dynamic and can be changed on the fly without reinstantiating several other redundant variables.

If we keep it in the config, we have to reinitialize the config with all the required arguments, like `serverConfig` because without it, this code breaks here:

```
 private static isValid(options: ConfigOptions): boolean {
    return !!(options && options.serverConfig);
  }
```

We can even remove loglevel from Config if we wanted and just keep it in the new instantiated singleton, i'm open to it either way. I just thought theres more of a chance a consumer will never call FRLogger and they are more likely to call Config.set, set the logging they want (or not) and never look at it again.

The more I wrote code the more I felt we wanted to keep these variables private so that warn / log / and error could be changed but they were never really leaked out of the implementation, so a closure was the only true way to do that.


I think theres more flexibility to an object approach, but we could i suppose, instantiate the class the same way, the token store doesn't accept a custom store so i don't see how similar it is in this sense, if it did the Token Store I don't think would be able to be abstract?

Reply:
Chatted on Slack. Let's see how the changes look.